ensime_shared/protocol_handler.py

```diff
@@ -0,0 +1,249 @@
+        raise NotImplementedError()
+
+    def show_type(self, call_id, payload):
+        raise NotImplementedError()
+
+
+class ProtocolHandlerV1(ProtocolHandler):
+
+    def __init__(self):
+        super(ProtocolHandlerV1, self).__init__()
+
+    def handle_indexer_ready(self, call_id, payload):
+        self.message("indexer_ready")
+
+    def handle_analyzer_ready(self, call_id, payload):
+        self.message("analyzer_ready")
+
+    def handle_debug_vm_error(self, call_id, payload):
+        self.vim.command(commands['display_message'].format("Error. Check ensime-vim log for details."))
+
+    def handle_import_suggestions(self, call_id, payload):
+        imports = list(sorted(set(suggestion['name'].replace('$', '.') for suggestions in payload['symLists'] for suggestion in suggestions)))
+        if imports:
+            chosen_import = int(self.vim.eval(commands['select_item_list'].format(json.dumps(
+                ["Select class to import:"] + ["{}. {}".format(num + 1, imp) for (num, imp) in enumerate(imports)]))))
+
+            if chosen_import > 0:
+                self.add_import(imports[chosen_import - 1])
+
+        else:
+            self.vim.command(commands['display_message'].format("No import suggestions found"))
+
+    def handle_package_info(self, call_id, payload):
+        package = payload["fullName"]
+        # Create a new buffer 45 columns wide
+        def add(member, indentLevel):
+            indent = "  " * indentLevel
+            t = member["declAs"]["typehint"] if member["typehint"] == "BasicTypeInfo" else ""
+            line = "{}{}: {}".format(indent, t, member["name"])
+            self.vim.command(commands["append_line"].format("\'$\'", str(line)))
+            if indentLevel < 4:
+                for m in member["members"]:
+                    add(m, indentLevel + 1)
+
+        cmd = commands["new_vertical_scratch"].format(str(45),"package_info")
+        self.vim.command(cmd)
+        self.vim.command(commands["set_filetype"].format("package_info"))
+        self.vim.command(commands["append_line"].format("\'$\'", str(package)))
+        for member in payload["members"]:
+            add(member, 1)
+
+    def handle_symbol_search(self, call_id, payload):
+        """Handler for symbol search results"""
+        self.log(payload)
+        syms = payload["syms"]
+        qfList = []
+        for sym in syms:
+            p = sym.get("pos")
+            if p:
+                item = self.to_quickfix_item(str(p["file"]),
+                                            p["line"],
+                                            str(sym["name"]),
+                                            "info")
+                qfList.append(item)
+        self.write_quickfix_list(qfList)
+
+    def handle_symbol_info(self, call_id, payload):
+        """Handler for response `SymbolInfo`."""
+        warn = lambda e: self.message("unknown_symbol")
+        with catch(KeyError, warn):
```

Comment:
Sorry to beat a dead horse and this isn't @ktonga's code so we're veering into the weeds, but this is a `with catch` with a 20-line body. I'm scared to refactor it because I'm not very sure where exactly the `KeyError`(s) could come from.

To be fair you could just as easily write a `try` block this big, but I believe the code smell of that is more obvious in code review. So the (slightly) greater syntactic weight of `try`/`catch` sometimes has merit. Here it might give you more encouragement to stop and wonder, "could I eliminate this error handling entirely by using `dict.get()`?".